File: PythonLeetcode/leetcodeM/116_PopulatingNextRightPointersInEachNode.py
"""
116. Populating Next Right Pointers in Each Node

You are given a perfect binary tree where all leaves are on the same level, and every parent has two children. The binary tree has the following definition:

struct Node {
  int val;
  Node *left;
  Node *right;
  Node *next;
}
Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.

Initially, all next pointers are set to NULL.

 

Follow up:

You may only use constant extra space.
Recursive approach is fine, you may assume implicit stack space does not count as extra space for this problem.
 

Example 1:



Input: root = [1,2,3,4,5,6,7]
Output: [1,#,2,3,#,4,5,6,7,#]
Explanation: Given the above perfect binary tree (Figure A), your function should populate each next pointer to point to its next right node, just like in Figure B. 
The serialized output is in level order as connected by the next pointers, with '#' signifying the end of each level.
 

Constraints:

The number of nodes in the given tree is less than 4096.
-1000 <= node.val <= 1000
"""


"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class PopulateNextRight:

    def doit_(self, root: 'Node') -> 'Node':
        from collections import deque
        
        if not root: return root
        
        st = deque([root])
        
        while st:
            
            prev = None
            size = len(st)

            for _ in range(size):
                
                node = st.popleft()
            
                if prev:
                    prev.next = node

                prev = node
                
                if node.left:
                    st.append(node.left)
                
                if node.right:
                    st.append(node.right)
        
        return root


    """
    Approach 1: Level Order Traversal
    Intuition

    There are two basic kinds of traversals on a tree or a graph. One is where we explore the tree in a depth first manner i.e. one branch at a time. 
    The other one is where we traverse the tree breadth-wise i.e. we explore one level of the tree before moving on to the next one. 
    For trees, we have further classifications of the depth first traversal approach called preorder, inorder, and the postorder traversals. 
    Breadth first approach to exploring a tree is based on the concept of the level of a node. The level of a node is its depth or distance from the root node. 
    We process all the nodes on one level before moving on to the next one.
    
    """
    def doit_(self, root: 'Node') -> 'Node':
        
        if not root:
            return root
        
        # Initialize a queue data structure which contains
        # just the root of the tree
        Q = collections.deque([root])
        
        # Outer while loop which iterates over 
        # each level
        while Q:
            
            # Note the size of the queue
            size = len(Q)
            
            # Iterate over all the nodes on the current level
            for i in range(size):
                
                # Pop a node from the front of the queue
                node = Q.popleft()
                
                # This check is important. We don't want to
                # establish any wrong connections. The queue will
                # contain nodes from 2 levels at most at any
                # point in time. This check ensures we only 
                # don't establish next pointers beyond the end
                # of a level
                if i < size - 1:
                    node.next = Q[0]
                
                # Add the children, if any, to the back of
                # the queue
                if node.left:
                    Q.append(node.left)
                if node.right:
                    Q.append(node.right)
        
        # Since the tree has now been modified, return the root node
        return root

    """
        Approach 2: Using previously established next pointers
        Intuition

        Let's look at the two types of next pointer connections we need to establish for a given tree.

        This first case is the one where we establish the next pointers between the two children of a given node. This is the easier of the two cases since both the children are accessible via the same node. 
        We can simply do the following to establish this connection.
    """
    def doit_self(self, root: 'Node') -> 'Node':
        
        if not root:
            return root
        
        # Start with the root node. There are no next pointers
        # that need to be set up on the first level
        leftmost = root
        
        # Once we reach the final level, we are done
        while leftmost.left:
            
            # Iterate the "linked list" starting from the head
            # node and using the next pointers, establish the 
            # corresponding links for the next level
            head = leftmost
            while head:
                
                # CONNECTION 1
                head.left.next = head.right
                
                # CONNECTION 2
                if head.next:
                    head.right.next = head.next.left
                
                # Progress along the list (nodes on the current level)
                head = head.next
            
            # Move onto the next level
            leftmost = leftmost.left
        
        return root

File: cppLeetcode/leetcodeM/116_PopulatingNextRightPointersInEachNode.cpp
/*
116. Populating Next Right Pointers in Each Node

You are given a perfect binary tree where all leaves are on the same level, and every parent has two children. The binary tree has the following definition:

struct Node {
  int val;
  Node *left;
  Node *right;
  Node *next;
}
Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.

Initially, all next pointers are set to NULL.

 

Follow up:

You may only use constant extra space.
Recursive approach is fine, you may assume implicit stack space does not count as extra space for this problem.
 

Example 1:



Input: root = [1,2,3,4,5,6,7]
Output: [1,#,2,3,#,4,5,6,7,#]
Explanation: Given the above perfect binary tree (Figure A), your function should populate each next pointer to point to its next right node, just like in Figure B. The serialized output is in level order as connected by the next pointers, with '#' signifying the end of each level.
 

Constraints:

The number of nodes in the given tree is less than 4096.
-1000 <= node.val <= 1000
*/

#include <functional>

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

#include <queue>
using std::queue;

class PopulatingNextRightPointers {



    Node* connect_best(Node* root) {
        
        Node *cur = root, *leftmost = NULL, *last = NULL;
        
        auto connect = [&](Node* child) {
            
            if (!child) return;
            
            if (!leftmost) {
                leftmost = child;
                last = child;
            } else {
                last->next = child;
                last = child;
            }
        };
        
        
        while (cur) {
            
            leftmost = NULL;
            last = NULL;
            
            
            while (cur) {
                
                connect(cur->left);
                connect(cur->right);
                
                cur = cur->next;
            }
            
            cur = leftmost;
        }
        
        return root;
    }

public:

    Node* connect(Node* root) {

        if (!root) return root;

        queue<Node*> buf;
        buf.push(root);


        while (!buf.empty()) {

            int size = buf.size();
            Node* last = nullptr;

            while(size--) {

                Node* cur = buf.front();
                buf.pop();

                if (last) last->next = cur;

                if (cur->left) buf.push(cur->left);
                if (cur->right) buf.push(cur->right);

                last = cur;
            }

        }

        return root;
    }
    
    /*
        Follow up:
        You may only use constant extra space.
        Recursive approach is fine, you may assume implicit stack space does not count as extra space for this problem.
    */
    Node* connect(Node* root) {
        
        std::function<void(Node*)> dfs = [&](Node* p) {
            
            if (!p) return;
            
            if (p->left) p->left->next = p->right;
            if (p->right) p->right->next = p->next ? p->next->left : NULL;
            
            dfs(p->left);
            dfs(p->right);
        };
        
        dfs(root);
        return root;
    }
};